## 1. Load Dataset

The DataCo Smart Supply Chain Dataset was loaded from the raw data directory using pandas.

The raw dataset is kept unchanged at this stage. No cleaning or transformation has been applied during the initial loading process.

In [213]:
import pandas as pd

In [214]:
file_path = "../data/raw/DataCoSupplyChainDataset.csv"

df = pd.read_csv(
    file_path,
    encoding="latin1"
)

In [215]:
print(f"Rows: {df.shape[0]:,}")
print(f"Columns: {df.shape[1]}")

Rows: 180,519
Columns: 53


In [216]:
df.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,...,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


## 2. Dataset Overview

The DataCo Smart Supply Chain Dataset contains 180,519 records and 53 variables covering order transactions, products, customers, sales, geography, shipping, and delivery information.

The dataset appears to be structured at the order-item level because multiple records may belong to the same Order Id while each record contains product and order-item attributes.

The dataset provides sufficient information to analyze order performance, sales performance, delivery efficiency, regional logistics performance, shipping modes, product categories, customer segments, and delivery delays.

However, warehouse performance, supplier performance, customer ratings, real-time shipment tracking, and route optimization are outside the scope because the required information is not available in the dataset or is beyond the scope of this historical analysis.

In [217]:
df.columns.tolist()

['Type',
 'Days for shipping (real)',
 'Days for shipment (scheduled)',
 'Benefit per order',
 'Sales per customer',
 'Delivery Status',
 'Late_delivery_risk',
 'Category Id',
 'Category Name',
 'Customer City',
 'Customer Country',
 'Customer Email',
 'Customer Fname',
 'Customer Id',
 'Customer Lname',
 'Customer Password',
 'Customer Segment',
 'Customer State',
 'Customer Street',
 'Customer Zipcode',
 'Department Id',
 'Department Name',
 'Latitude',
 'Longitude',
 'Market',
 'Order City',
 'Order Country',
 'Order Customer Id',
 'order date (DateOrders)',
 'Order Id',
 'Order Item Cardprod Id',
 'Order Item Discount',
 'Order Item Discount Rate',
 'Order Item Id',
 'Order Item Product Price',
 'Order Item Profit Ratio',
 'Order Item Quantity',
 'Sales',
 'Order Item Total',
 'Order Profit Per Order',
 'Order Region',
 'Order State',
 'Order Status',
 'Order Zipcode',
 'Product Card Id',
 'Product Category Id',
 'Product Description',
 'Product Image',
 'Product Name',
 'Product P

In [218]:
numeric_columns = df.select_dtypes(
    include=["int64", "float64"]
).columns.tolist()

print(f"Number of numerical columns: {len(numeric_columns)}")
print(numeric_columns)

Number of numerical columns: 29
['Days for shipping (real)', 'Days for shipment (scheduled)', 'Benefit per order', 'Sales per customer', 'Late_delivery_risk', 'Category Id', 'Customer Id', 'Customer Zipcode', 'Department Id', 'Latitude', 'Longitude', 'Order Customer Id', 'Order Id', 'Order Item Cardprod Id', 'Order Item Discount', 'Order Item Discount Rate', 'Order Item Id', 'Order Item Product Price', 'Order Item Profit Ratio', 'Order Item Quantity', 'Sales', 'Order Item Total', 'Order Profit Per Order', 'Order Zipcode', 'Product Card Id', 'Product Category Id', 'Product Description', 'Product Price', 'Product Status']


In [219]:
categorical_columns = df.select_dtypes(
    include=["object", "str"]
).columns.tolist()

print(f"Number of categorical/text columns: {len(categorical_columns)}")
print(categorical_columns)

Number of categorical/text columns: 24
['Type', 'Delivery Status', 'Category Name', 'Customer City', 'Customer Country', 'Customer Email', 'Customer Fname', 'Customer Lname', 'Customer Password', 'Customer Segment', 'Customer State', 'Customer Street', 'Department Name', 'Market', 'Order City', 'Order Country', 'order date (DateOrders)', 'Order Region', 'Order State', 'Order Status', 'Product Image', 'Product Name', 'shipping date (DateOrders)', 'Shipping Mode']


## 3. Data Types

This section examines the data types of each variable to identify numerical, categorical, identifier, and date-related fields.

At this stage, no data type conversion is performed. Any required conversion will be addressed during the data cleaning and preparation phase.

In [220]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  str    
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  str    
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  str    
 9   Customer City                  180519 non-null  str    
 10  Customer Country               180519 non-null  str    
 11  Customer Email                 180519 non-null  str    
 12  Customer Fname                 180519 non

In [221]:
dtype_summary = pd.DataFrame({
    "columns": df.columns,
    "dtype": df.dtypes.astype(str).values
})

dtype_summary

,columns,dtype
0,Type,str
1,Days for shipping (real),int64
2,Days for shipment (scheduled),int64
3,Benefit per order,float64
4,Sales per customer,float64
5,Delivery Status,str
6,Late_delivery_risk,int64
7,Category Id,int64
8,Category Name,str
9,Customer City,str


In [222]:
df.dtypes.value_counts()

str        24
float64    15
int64      14
Name: count, dtype: int64

In [223]:
date_columns = [
    "order_date (DateOrders)",
    "shipping_date (DateOrders)",
]

print("Potential date columns:")
print("\n".join(date_columns))

Potential date columns:
order_date (DateOrders)
shipping_date (DateOrders)


In [224]:
identifier_columns = [
    "Order Id",
    "Order Item Id",
    "Customer Id",
    "Order Customer Id",
    "Product Card Id",
    "Category Id",
    "Product Category Id",
    "Department Id"
]

print("Potential identifier columns:")
print("\n".join(identifier_columns))

Potential identifier columns:
Order Id
Order Item Id
Customer Id
Order Customer Id
Product Card Id
Category Id
Product Category Id
Department Id


### Data Type Assessment

The dataset contains 53 variables consisting of 29 numerical columns and 24 text/string columns.

Most numerical variables have appropriate data types for quantitative analysis. However, several integer columns such as Order Id and Customer Id function primarily as identifiers rather than analytical measures.

Two date-related columns, order date (DateOrders) and shipping date (DateOrders), are currently stored as string values even though they represent date/time information. These columns will require conversion during the data cleaning and preparation phase.

Binary and status-related variables such as Late_delivery_risk and Product Status are stored as integers and should be treated as categorical or flag variables when appropriate during analysis.

Product Description is currently stored as float64 even though it is expected to contain textual product descriptions. This column will be investigated further during the missing value and data quality assessment stages.

## 4. Missing Values

This section evaluates the completeness of the dataset by identifying missing values in each column.

No missing values are handled or removed at this stage. The purpose is to understand the extent and distribution of missing data before determining appropriate treatment during the data cleaning phase.

In [225]:
missing_summary = pd.DataFrame({
    "missing_count": df.isnull().sum(),
    "missing_percentage": (df.isnull().sum() / len(df)) * 100
})

missing_summary = missing_summary.sort_values(
    by="missing_count",
    ascending=False
)

missing_summary

,missing_count,missing_percentage
Product Description,180519,100.000000
Order Zipcode,155679,86.239676
Customer Lname,8,0.004432
Customer Zipcode,3,0.001662
Days for shipment (scheduled),0,0.000000
Sales per customer,0,0.000000
Benefit per order,0,0.000000
Delivery Status,0,0.000000
Late_delivery_risk,0,0.000000
Customer City,0,0.000000


In [226]:
missing_columns = missing_summary[
    missing_summary["missing_count"] > 0
]

missing_columns

,missing_count,missing_percentage
Product Description,180519,100.000000
Order Zipcode,155679,86.239676
Customer Lname,8,0.004432
Customer Zipcode,3,0.001662


In [229]:
fully_missing_columns = missing_summary[
    missing_summary["missing_percentage"] == 100
]

fully_missing_columns

,missing_count,missing_percentage
Product Description,180519,100.0


In [231]:
def classify_missing_percentage(x):
    if x == 0:
        return "No Missing"
    elif x < 5:
        return "< 5%"
    elif x < 20:
        return "5% - <20%"
    elif x < 50:
        return "20% - <50%"
    elif x < 100:
        return "50% - <100%"
    else:
        return "100%"

missing_summary["missing_category"] = (
    missing_summary["missing_percentage"]
    .apply(classify_missing_percentage)
)

missing_summary

,missing_count,missing_percentage,missing_category
Product Description,180519,100.000000,100%
Order Zipcode,155679,86.239676,50% - <100%
Customer Lname,8,0.004432,< 5%
Customer Zipcode,3,0.001662,< 5%
Days for shipment (scheduled),0,0.000000,No Missing
Sales per customer,0,0.000000,No Missing
Benefit per order,0,0.000000,No Missing
Delivery Status,0,0.000000,No Missing
Late_delivery_risk,0,0.000000,No Missing
Customer City,0,0.000000,No Missing


In [232]:
missing_summary["missing_category"].value_counts()

missing_category
No Missing     49
< 5%            2
100%            1
50% - <100%     1
Name: count, dtype: int64

### Missing Value Assessment

The dataset contains missing values in only 4 of the 53 variables.

Product Description is completely missing across all 180,519 records, making the column unavailable for analytical use.

Order Zipcode has a high missing rate of 86.24%, making it unsuitable as the primary geographic dimension. Geographic analysis should instead rely on more complete fields such as Order Region, Order Country, Order State, and Order City.

Customer Lname and Customer Zipcode contain only a very small number of missing records, representing less than 0.01% of the dataset. These missing values are unlikely to materially affect the core supply chain and logistics analysis.

No missing values are removed or imputed at this stage. Treatment decisions will be made during the data cleaning and preparation phase based on analytical relevance and business requirements.

## 5. Duplicates

This section evaluates duplicate records in the dataset.

Duplicate analysis is performed at multiple levels because the dataset appears to be structured at the order-item level. Repeated Order Id values do not necessarily indicate duplicate records, as a single order may contain multiple items.

No records are removed at this stage. The purpose is to distinguish true duplicate records from legitimate repeated business transactions.

In [235]:
duplicate_count = df.duplicated().sum()

print(f"Exact duplicate rows: {duplicate_count:,}")

Exact duplicate rows: 0


In [236]:
duplicate_percentage = (
    duplicate_count / len(df)
) * 100

print(f"Duplicate percentage: {duplicate_percentage:.2f}%")

Duplicate percentage: 0.00%


In [303]:
order_id_counts = df.groupby("Order Id").size()
order_item_id_counts = df.groupby("Order Item Id").size()

In [307]:
duplicate_summary = pd.DataFrame({
    "Metric": [
        "Total Rows",
        "Unique Order Ids",
        "Unique Order Item Ids",
        "Exact Duplicate Rows",
        "Orders with Multiple Rows",
        "Order Item Ids Appearing More Than Once"
    ],
    "Count": [
        len(df),
        df["Order Id"].nunique(),
        df["Order Item Id"].nunique(),
        duplicate_count,
        (order_id_counts > 1).sum(),
        (order_item_id_counts > 1).sum()
    ]
})

duplicate_summary

,Metric,Count
0,Total Rows,180519
1,Unique Order Ids,65752
2,Unique Order Item Ids,180519
3,Exact Duplicate Rows,0
4,Orders with Multiple Rows,45902
5,Order Item Ids Appearing More Than Once,0


In [308]:
order_id_counts.describe()

count    65752.000000
mean         2.745453
std          1.478363
min          1.000000
25%          1.000000
50%          3.000000
75%          4.000000
max          5.000000
dtype: float64

In [242]:
multi_row_orders = order_id_counts[
    order_id_counts > 1
]

multi_row_orders.head(10)

Order Id
2     3
4     4
5     5
7     3
8     4
9     3
10    5
11    5
12    5
14    3
dtype: int64

In [243]:
example_order_id = multi_row_orders.index[0]

df[df["Order Id"] == example_order_id][[
    "Order Id",
    "Order Item Id",
    "Product Name",
    "Order Item Quantity",
    "Sales",
    "Order Item Total"
]]

,Order Id,Order Item Id,Product Name,Order Item Quantity,Sales,Order Item Total
77011,2,2,Pelican Sunstream 100 Kayak,1,199.990005,193.990005
87884,2,4,Nike Men's CJ Elite 2 TD Football Cleat,1,129.990005,107.889999
109322,2,3,Nike Men's Dri-FIT Victory Golf Polo,5,250.000000,227.500000


### Duplicate and Dataset Grain Assessment

The dataset contains 180,519 records with no exact duplicate rows. The exact duplicate rate is 0.00%, indicating that there are no fully duplicated records in the raw dataset.

The dataset contains 65,752 unique Order Ids, while 45,902 orders consist of multiple rows. The number of rows per order ranges from 1 to 5, with an average of approximately 2.75 rows per order.

Each Order Item Id appears exactly once, with 180,519 unique Order Item Ids and no repeated identifiers.

These findings confirm that the dataset is structured at the order-item level, where each row represents one order item. Therefore, the number of rows should not be interpreted as the total number of orders.

For order-level metrics, Order Id must be used as the unique identifier to avoid double counting.

## 6. Unique Values

This section examines the number and distribution of unique values across the dataset.

The analysis focuses on identifier columns, categorical dimensions, and binary or status variables to understand their cardinality and validate whether the observed values are reasonable from a business perspective.

No values are modified or removed at this stage.

In [249]:
unique_summary = pd.DataFrame({
    "unique_count": df.nunique(),
    "unique_percentage": (df.nunique() / len(df)) * 100
})

unique_summary = unique_summary.sort_values(
    by="unique_count",
    ascending=False
)

unique_summary["unique_percentage"] = (
    unique_summary["unique_percentage"].round(2)
)

unique_summary

,unique_count,unique_percentage
Order Item Id,180519,100.00
order date (DateOrders),65752,36.42
Order Id,65752,36.42
shipping date (DateOrders),63701,35.29
Benefit per order,21998,12.19
Order Profit Per Order,21998,12.19
Customer Id,20652,11.44
Order Customer Id,20652,11.44
Latitude,11250,6.23
Customer Street,7458,4.13


In [250]:
identifier_columns = [
    "Order Id",
    "Order Item Id",
    "Customer Id",
    "Order Customer Id",
    "Product Card Id",
    "Product Category Id",
    "Category Id",
    "Department Id"
]

identifier_summary = pd.DataFrame({
    "unique_count": df[identifier_columns].nunique(),
    "total_rows": len(df),
    "unique_percentage": (
        df[identifier_columns].nunique() / len(df) * 100
    ).round(2)
})

identifier_summary

,unique_count,total_rows,unique_percentage
Order Id,65752,180519,36.42
Order Item Id,180519,180519,100.00
Customer Id,20652,180519,11.44
Order Customer Id,20652,180519,11.44
Product Card Id,118,180519,0.07
Product Category Id,51,180519,0.03
Category Id,51,180519,0.03
Department Id,11,180519,0.01


In [251]:
print("Customer Id unique values:",
      df["Customer Id"].nunique())

print("Order Customer Id unique values:",
      df["Order Customer Id"].nunique())

print(
    "Customer Id and Order Customer Id have the same values:",
    set(df["Customer Id"].unique()) == set(df["Order Customer Id"].unique())
)

Customer Id unique values: 20652
Order Customer Id unique values: 20652
Customer Id and Order Customer Id have the same values: True


In [252]:
categorical_columns = [
    "Type",
    "Delivery Status",
    "Category Name",
    "Customer Country",
    "Customer Segment",
    "Customer State",
    "Department Name",
    "Market",
    "Order City",
    "Order Country",
    "Order Region",
    "Order State",
    "Order Status",
    "Product Name",
    "Shipping Mode"
]

In [253]:
categorical_summary = pd.DataFrame({
    "unique_count": df[categorical_columns].nunique(),
    "unique_percentage": (
        df[categorical_columns].nunique() / len(df) * 100
    ).round(2)
})

categorical_summary

,unique_count,unique_percentage
Type,4,0.00
Delivery Status,4,0.00
Category Name,50,0.03
Customer Country,2,0.00
Customer Segment,3,0.00
Customer State,46,0.03
Department Name,11,0.01
Market,5,0.00
Order City,3597,1.99
Order Country,164,0.09


In [254]:
df["Shipping Mode"].unique()

<StringArray>
['Standard Class', 'First Class', 'Second Class', 'Same Day']
Length: 4, dtype: str

In [255]:
df["Delivery Status"].unique()

<StringArray>
['Advance shipping', 'Late delivery', 'Shipping on time', 'Shipping canceled']
Length: 4, dtype: str

In [256]:
df["Order Status"].unique()

<StringArray>
[       'COMPLETE',         'PENDING',          'CLOSED', 'PENDING_PAYMENT',
        'CANCELED',      'PROCESSING', 'SUSPECTED_FRAUD',         'ON_HOLD',
  'PAYMENT_REVIEW']
Length: 9, dtype: str

In [257]:
df["Customer Segment"].unique()

<StringArray>
['Consumer', 'Home Office', 'Corporate']
Length: 3, dtype: str

In [258]:
df["Market"].unique()

<StringArray>
['Pacific Asia', 'USCA', 'Africa', 'Europe', 'LATAM']
Length: 5, dtype: str

In [259]:
df["Type"].unique()

<StringArray>
['DEBIT', 'TRANSFER', 'CASH', 'PAYMENT']
Length: 4, dtype: str

In [260]:
regional_columns = [
    "Market",
    "Order Region",
    "Order Country",
    "Order State",
    "Order City"
]

regional_summary = pd.DataFrame({
    "unique_count": df[regional_columns].nunique()
})

regional_summary

,unique_count
Market,5
Order Region,23
Order Country,164
Order State,1089
Order City,3597


In [261]:
for column in regional_columns:
    print(f"\n{column}")
    print(df[column].unique()[:20])


Market
<StringArray>
['Pacific Asia', 'USCA', 'Africa', 'Europe', 'LATAM']
Length: 5, dtype: str

Order Region
<StringArray>
[ 'Southeast Asia',      'South Asia',         'Oceania',    'Eastern Asia',
       'West Asia',    'West of USA ',      'US Center ',     'West Africa',
  'Central Africa',    'North Africa',  'Western Europe', 'Northern Europe',
 'Central America',       'Caribbean',   'South America',     'East Africa',
 'Southern Europe',     'East of USA',          'Canada', 'Southern Africa']
Length: 20, dtype: str

Order Country
<StringArray>
[                      'Indonesia',                           'India',
                       'Australia',                           'China',
                           'Japón',                   'Corea del Sur',
                        'Singapur',                         'Turquía',
                        'Mongolia',                  'Estados Unidos',
                         'Nigeria', 'República Democrática del Congo',
           

In [262]:
product_columns = [
    "Category Name",
    "Product Name",
    "Product Card Id",
    "Product Category Id",
    "Department Name",
    "Department Id"
]

product_summary = pd.DataFrame({
    "unique_count": df[product_columns].nunique(),
    "unique_percentage": (
        df[product_columns].nunique() / len(df) * 100
    ).round(2)
})

product_summary

,unique_count,unique_percentage
Category Name,50,0.03
Product Name,118,0.07
Product Card Id,118,0.07
Product Category Id,51,0.03
Department Name,11,0.01
Department Id,11,0.01


In [264]:
df["Late_delivery_risk"].unique()

array([0, 1])

In [265]:
df["Product Status"].unique()

array([0])

In [266]:
df["Late_delivery_risk"].value_counts()

Late_delivery_risk
1    98977
0    81542
Name: count, dtype: int64

In [267]:
df["Product Status"].value_counts()

Product Status
0    180519
Name: count, dtype: int64

In [269]:
df["Customer Country"].value_counts(dropna=False)

Customer Country
EE. UU.        111146
Puerto Rico     69373
Name: count, dtype: int64

In [270]:
category_mapping = (
    df[["Category Id", "Category Name"]]
    .drop_duplicates()
    .sort_values("Category Id")
)

category_mapping

,Category Id,Category Name
175,2,Soccer
119,3,Baseball & Softball
174,4,Basketball
775,5,Lacrosse
294,6,Tennis & Racquet
2569,7,Hockey
59,9,Cardio Equipment
547,10,Strength Training
301,11,Fitness Accessories
56,12,Boxing & MMA


In [271]:
category_id_check = (
    df.groupby("Category Id")["Category Name"]
    .nunique()
    .sort_values(ascending=False)
)

category_id_check

Category Id
2     1
3     1
4     1
5     1
6     1
7     1
9     1
10    1
11    1
12    1
13    1
16    1
17    1
18    1
24    1
26    1
29    1
30    1
31    1
32    1
33    1
34    1
35    1
36    1
37    1
38    1
40    1
41    1
43    1
44    1
45    1
46    1
48    1
59    1
60    1
61    1
62    1
63    1
64    1
65    1
66    1
67    1
68    1
69    1
70    1
71    1
72    1
73    1
74    1
75    1
76    1
Name: Category Name, dtype: int64

In [272]:
category_name_check = (
    df.groupby("Category Name")["Category Id"]
    .nunique()
    .sort_values(ascending=False)
)

category_name_check

Category Name
Electronics             2
Accessories             1
As Seen on  TV!         1
Baby                    1
Basketball              1
Baseball & Softball     1
Boxing & MMA            1
CDs                     1
Cameras                 1
Books                   1
Camping & Hiking        1
Cardio Equipment        1
Cleats                  1
Children's Clothing     1
Computers               1
Consumer Electronics    1
Crafts                  1
DVDs                    1
Fishing                 1
Fitness Accessories     1
Garden                  1
Girls' Apparel          1
Golf Apparel            1
Golf Bags & Carts       1
Golf Balls              1
Golf Gloves             1
Golf Shoes              1
Health and Beauty       1
Hockey                  1
Hunting & Shooting      1
Indoor/Outdoor Games    1
Kids' Golf Clubs        1
Lacrosse                1
Men's Clothing          1
Men's Footwear          1
Men's Golf Clubs        1
Music                   1
Pet Supplies            

In [273]:
product_mapping_check = (
    df.groupby("Product Card Id")["Product Name"]
    .nunique()
)

product_mapping_check.value_counts()

Product Name
1    118
Name: count, dtype: int64

### Unique Values Assessment

The unique-value analysis confirms that the dataset contains a clear hierarchy of identifiers, transactional dimensions, geographic dimensions, product attributes, and operational status variables.

The dataset contains 180,519 unique Order Item Ids and 65,752 unique Order Ids, confirming the previously identified order-item grain. There are also 20,652 unique customers and 118 unique products.

The main categorical variables have reasonable cardinality for supply chain and logistics analysis, including 4 shipping modes, 4 delivery statuses, 9 order statuses, 5 markets, and 23 order regions.

The geographic dimensions show increasing cardinality from Market to Order Region, Order Country, Order State, and Order City, providing sufficient granularity for regional logistics analysis.

The Category Id to Category Name relationship is consistent, with each Category Id mapping to a single Category Name. However, the Category Name "Electronics" is associated with two Category Ids, indicating that category names are not always unique identifiers.

Several columns have no analytical variation. Customer Email, Customer Password, and Product Status contain only one unique value, while Product Description is completely missing. These columns will be evaluated for removal during the data cleaning and preparation phase.

The Late_delivery_risk column contains two values (0 and 1), making it a valid binary indicator. However, it will not be directly interpreted as the project's final delay rate because delivery performance metrics will be derived separately during the analytical phase.

No unexpected categorical values or inconsistent identifier mappings were identified during this assessment.

## 7. Outlier Analysis

Outlier analysis is performed to identify unusually high or low observations in numerical variables that may require further investigation.

Not all numerical columns are suitable for outlier detection. Identifier columns, binary indicators, categorical codes, and constant variables are excluded because extreme values in these fields do not represent meaningful business outliers.

The analysis focuses on business-related numerical variables such as sales, profit, discounts, quantities, prices, delivery time, and customer-related operational measures.

Statistical outliers are not automatically treated as errors. Extreme observations may represent legitimate business transactions, high-value orders, large quantities, or unusual operational conditions. Any treatment of outliers will therefore be determined during the data cleaning phase based on both statistical evidence and business context.

In [274]:
outlier_columns = [
    "Days for shipping (real)",
    "Days for shipment (scheduled)",
    "Benefit per order",
    "Sales per customer",
    "Order Item Discount",
    "Order Item Discount Rate",
    "Order Item Product Price",
    "Order Item Profit Ratio",
    "Order Item Quantity",
    "Sales",
    "Order Item Total",
    "Order Profit Per Order",
    "Product Price"
]

In [275]:
outlier_summary = df[outlier_columns].describe().T

outlier_summary

,count,mean,std,min,25%,50%,75%,max
Days for shipping (real),180519.0,3.497654,1.623722,0.00000,2.000000,3.000000,5.000000,6.000000
Days for shipment (scheduled),180519.0,2.931847,1.374449,0.00000,2.000000,4.000000,4.000000,4.000000
Benefit per order,180519.0,21.974989,104.433526,-4274.97998,7.000000,31.520000,64.800003,911.799988
Sales per customer,180519.0,183.107609,120.043670,7.49000,104.379997,163.990005,247.399994,1939.989990
Order Item Discount,180519.0,20.664741,21.800901,0.00000,5.400000,14.000000,29.990000,500.000000
Order Item Discount Rate,180519.0,0.101668,0.070415,0.00000,0.040000,0.100000,0.160000,0.250000
Order Item Product Price,180519.0,141.232550,139.732492,9.99000,50.000000,59.990002,199.990005,1999.989990
Order Item Profit Ratio,180519.0,0.120647,0.466796,-2.75000,0.080000,0.270000,0.360000,0.500000
Order Item Quantity,180519.0,2.127638,1.453451,1.00000,1.000000,1.000000,3.000000,5.000000
Sales,180519.0,203.772096,132.273077,9.99000,119.980003,199.919998,299.950012,1999.989990


In [276]:
Q1 = df[outlier_columns].quantile(0.25)
Q3 = df[outlier_columns].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

In [277]:
outlier_count = {}

for column in outlier_columns:
    outlier_count[column] = (
        (df[column] < lower_bound[column]) |
        (df[column] > upper_bound[column])
    ).sum()

outlier_summary_iqr = pd.DataFrame({
    "outlier_count": outlier_count
})

outlier_summary_iqr["outlier_percentage"] = (
    outlier_summary_iqr["outlier_count"] / len(df) * 100
).round(2)

outlier_summary_iqr = outlier_summary_iqr.sort_values(
    "outlier_count",
    ascending=False
)

outlier_summary_iqr

,outlier_count,outlier_percentage
Benefit per order,18942,10.49
Order Profit Per Order,18942,10.49
Order Item Profit Ratio,17300,9.58
Order Item Discount,7537,4.18
Product Price,2048,1.13
Order Item Product Price,2048,1.13
Order Item Total,1943,1.08
Sales per customer,1943,1.08
Sales,488,0.27
Days for shipment (scheduled),0,0.00


In [278]:
outlier_analysis = pd.DataFrame({
    "Q1": Q1,
    "Q3": Q3,
    "IQR": IQR,
    "Lower Bound": lower_bound,
    "Upper Bound": upper_bound,
    "Outlier Count": [
        outlier_count[column]
        for column in outlier_columns
    ]
})

outlier_analysis["Outlier Percentage"] = (
    outlier_analysis["Outlier Count"] / len(df) * 100
).round(2)

outlier_analysis

,Q1,Q3,IQR,Lower Bound,Upper Bound,Outlier Count,Outlier Percentage
Days for shipping (real),2.000000,5.000000,3.000000,-2.500000,9.500000,0,0.00
Days for shipment (scheduled),2.000000,4.000000,2.000000,-1.000000,7.000000,0,0.00
Benefit per order,7.000000,64.800003,57.800003,-79.700005,151.500008,18942,10.49
Sales per customer,104.379997,247.399994,143.019997,-110.149998,461.929989,1943,1.08
Order Item Discount,5.400000,29.990000,24.590000,-31.484999,66.874999,7537,4.18
Order Item Discount Rate,0.040000,0.160000,0.120000,-0.140000,0.340000,0,0.00
Order Item Product Price,50.000000,199.990005,149.990005,-174.985008,424.975014,2048,1.13
Order Item Profit Ratio,0.080000,0.360000,0.280000,-0.340000,0.780000,17300,9.58
Order Item Quantity,1.000000,3.000000,2.000000,-2.000000,6.000000,0,0.00
Sales,119.980003,299.950012,179.970009,-149.975010,569.905025,488,0.27


In [279]:
df["Order Item Quantity"].value_counts().sort_index()

Order Item Quantity
1    99134
2    20315
3    20350
4    20335
5    20385
Name: count, dtype: int64

In [280]:
df[
    [
        "Days for shipping (real)",
        "Days for shipment (scheduled)"
    ]
].describe()

,Days for shipping (real),Days for shipment (scheduled)
count,180519.000000,180519.000000
mean,3.497654,2.931847
std,1.623722,1.374449
min,0.000000,0.000000
25%,2.000000,2.000000
50%,3.000000,4.000000
75%,5.000000,4.000000
max,6.000000,4.000000


In [281]:
df["Days for shipping (real)"].value_counts().sort_index()

Days for shipping (real)
0     5080
1     4657
2    56618
3    28765
4    28513
5    28163
6    28723
Name: count, dtype: int64

In [282]:
df["Days for shipment (scheduled)"].value_counts().sort_index()

Days for shipment (scheduled)
0      9737
1     27814
2     35216
4    107752
Name: count, dtype: int64

In [283]:
df[
    [
        "Sales",
        "Benefit per order",
        "Order Profit Per Order",
        "Order Item Total",
        "Product Price"
    ]
].describe().T

,count,mean,std,min,25%,50%,75%,max
Sales,180519.0,203.772096,132.273077,9.99000,119.980003,199.919998,299.950012,1999.989990
Benefit per order,180519.0,21.974989,104.433526,-4274.97998,7.000000,31.520000,64.800003,911.799988
Order Profit Per Order,180519.0,21.974989,104.433526,-4274.97998,7.000000,31.520000,64.800003,911.799988
Order Item Total,180519.0,183.107609,120.043670,7.49000,104.379997,163.990005,247.399994,1939.989990
Product Price,180519.0,141.232550,139.732492,9.99000,50.000000,59.990002,199.990005,1999.989990


### Outlier Assessment

The IQR method was used to identify statistical outliers in business-related numerical variables. Identifier columns, categorical codes, binary indicators, and constant variables were excluded from this analysis.

Delivery-related variables showed no statistical outliers. Days for shipping (real) ranges from 0 to 6 days, while Days for shipment (scheduled) ranges from 0 to 4 days. Order Item Quantity ranges from 1 to 5 units and also contains no IQR-based outliers.

Several financial and commercial variables contain statistical outliers. Sales has 488 outliers (0.27%), Product Price has 2,048 outliers (1.13%), and Order Item Discount has 7,537 outliers (4.18%). Profit-related variables show a higher number of statistical outliers, particularly Benefit per order and Order Profit Per Order with 18,942 observations (10.49%).

These observations are not automatically considered data quality issues. High sales, product prices, discounts, and profits may represent legitimate high-value transactions, while negative profit values may represent genuine loss-making transactions.

Therefore, no observations are removed at this stage based solely on the IQR method. Potentially extreme financial and operational observations will be retained and further evaluated during the data cleaning and analytical phases using business rules and domain context.

## 8. Data Quality

In [284]:
# Convert date columns temporarily for validation
order_date = pd.to_datetime(
    df["order date (DateOrders)"],
    errors="coerce"
)

shipping_date = pd.to_datetime(
    df["shipping date (DateOrders)"],
    errors="coerce"
)

invalid_date_order = (
    shipping_date < order_date
).sum()

In [285]:
negative_real_shipping = (
    df["Days for shipping (real)"] < 0
).sum()

negative_scheduled_shipping = (
    df["Days for shipment (scheduled)"] < 0
).sum()

In [286]:
invalid_quantity = (
    df["Order Item Quantity"] <= 0
).sum()

In [309]:
invalid_discount = (
    df["Order Item Discount"] < 0
).sum()

In [310]:
invalid_discount_rate = (
    (df["Order Item Discount Rate"] < 0) |
    (df["Order Item Discount Rate"] > 1)
).sum()

In [311]:
expected_delivery_status = {
    "Advance shipping",
    "Late delivery",
    "Shipping on time",
    "Shipping canceled"
}

invalid_delivery_status = (
    ~df["Delivery Status"].isin(expected_delivery_status)
).sum()

In [312]:
late_status = (
    df["Delivery Status"] == "Late delivery"
)

risk_flag = (
    df["Late_delivery_risk"] == 1
)

mismatch_count = (
    late_status != risk_flag
).sum()

In [313]:
invalid_product_status = (
    ~df["Product Status"].isin([0, 1])
).sum()

In [314]:
customer_id_mismatch = (
    df["Customer Id"] != df["Order Customer Id"]
).sum()

In [315]:
duplicate_order_item_id = (
    df["Order Item Id"].duplicated()
).sum()

In [316]:
quality_checks = {
    "Invalid order dates": order_date.isna().sum(),
    "Invalid shipping dates": shipping_date.isna().sum(),
    "Shipping date before order date": invalid_date_order,
    "Negative actual shipping days": negative_real_shipping,
    "Negative scheduled shipping days": negative_scheduled_shipping,
    "Invalid order quantities": invalid_quantity,
    "Negative discounts": invalid_discount,
    "Invalid discount rates": invalid_discount_rate,
    "Invalid delivery status": invalid_delivery_status,
    "Late delivery risk mismatches": mismatch_count,
    "Invalid product status": invalid_product_status,
    "Customer ID mismatches": customer_id_mismatch,
    "Duplicate Order Item IDs": duplicate_order_item_id
}

quality_summary = pd.DataFrame(
    quality_checks.items(),
    columns=["Check", "Issue Count"]
)

quality_summary

,Check,Issue Count
0,Invalid order dates,0
1,Invalid shipping dates,0
2,Shipping date before order date,0
3,Negative actual shipping days,0
4,Negative scheduled shipping days,0
5,Invalid order quantities,0
6,Negative discounts,0
7,Invalid discount rates,0
8,Invalid delivery status,0
9,Late delivery risk mismatches,0


In [287]:
price_columns = [
    "Order Item Product Price",
    "Product Price",
    "Sales",
    "Order Item Total",
    "Sales per customer"
]

for col in price_columns:
    invalid_count = (df[col] <= 0).sum()
    print(f"{col}: {invalid_count} invalid values")

Order Item Product Price: 0 invalid values
Product Price: 0 invalid values
Sales: 0 invalid values
Order Item Total: 0 invalid values
Sales per customer: 0 invalid values


In [290]:
profit_ratio = df["Order Item Profit Ratio"]

print("Minimum profit ratio:", profit_ratio.min())
print("Maximum profit ratio:", profit_ratio.max())

Minimum profit ratio: -2.75
Maximum profit ratio: 0.5


In [292]:
delivery_status_counts = (
    df["Delivery Status"]
    .value_counts()
)

print(delivery_status_counts)

Delivery Status
Late delivery        98977
Advance shipping     41592
Shipping on time     32196
Shipping canceled     7754
Name: count, dtype: int64


In [294]:
risk_status_check = pd.crosstab(
    df["Delivery Status"],
    df["Late_delivery_risk"]
)

risk_status_check

Late_delivery_risk,0,1
Delivery Status,,
Advance shipping,41592,0
Late delivery,0,98977
Shipping canceled,7754,0
Shipping on time,32196,0


### Data Quality Assessment

A set of business-rule validation checks was performed to evaluate the structural and logical quality of the dataset.

The validation covered date validity, shipping time consistency, order quantity, discount values, delivery status, late delivery risk, product status, customer identifiers, and order item identifiers.

All business-rule validation checks returned zero issues. No invalid dates, negative shipping durations, invalid quantities, negative discounts, invalid discount rates, invalid delivery statuses, inconsistent customer identifiers, or duplicate order item identifiers were detected.

The dataset also contains several statistical outliers identified during the previous outlier analysis. However, these observations are not automatically considered data quality issues because extreme sales, prices, discounts, and negative profit values can represent legitimate business transactions.

Therefore, the dataset is considered structurally and logically consistent based on the business rules evaluated at this stage. No records will be removed based solely on these validation results.

Further data transformation and cleaning will be performed during Phase 3, where data types, missing values, date fields, redundant columns, and analytical features will be prepared for downstream analysis.

## 9. Data Dictionary

In [300]:
data_dictionary = [
    ["Type", "str", "Dimension", "Payment transaction type", "Payment analysis"],
    ["Days for shipping (real)", "int64", "Measure", "Actual shipping duration in days", "Delivery performance"],
    ["Days for shipment (scheduled)", "int64", "Measure", "Scheduled shipping duration in days", "Delivery benchmark"],
    ["Benefit per order", "float64", "Measure", "Benefit generated per order", "Profitability analysis"],
    ["Sales per customer", "float64", "Measure", "Sales value associated with the customer/order", "Customer value analysis"],
    ["Delivery Status", "str", "Dimension", "Delivery outcome or status", "Delivery performance"],
    ["Late_delivery_risk", "int64", "Flag", "Indicator of late delivery risk", "Delay analysis"],
    ["Category Id", "int64", "Identifier", "Product category identifier", "Category analysis"],
    ["Category Name", "str", "Dimension", "Product category name", "Category analysis"],
    ["Customer City", "str", "Dimension", "Customer city", "Customer geography"],
    ["Customer Country", "str", "Dimension", "Customer country", "Customer geography"],
    ["Customer Email", "str", "Metadata", "Customer email address", "Not required for core analysis"],
    ["Customer Fname", "str", "Metadata", "Customer first name", "Not required for core analysis"],
    ["Customer Id", "int64", "Identifier", "Customer identifier", "Customer analysis"],
    ["Customer Lname", "str", "Metadata", "Customer last name", "Not required for core analysis"],
    ["Customer Password", "str", "Metadata", "Customer password field", "Exclude from analytical dataset"],
    ["Customer Segment", "str", "Dimension", "Customer segment classification", "Customer analysis"],
    ["Customer State", "str", "Dimension", "Customer state", "Customer geography"],
    ["Customer Street", "str", "Metadata", "Customer street address", "Not required for core analysis"],
    ["Customer Zipcode", "float64", "Dimension", "Customer postal code", "Customer geography"],
    ["Department Id", "int64", "Identifier", "Department identifier", "Department analysis"],
    ["Department Name", "str", "Dimension", "Department name", "Department analysis"],
    ["Latitude", "float64", "Measure", "Geographic latitude", "Geographic and distance analysis"],
    ["Longitude", "float64", "Measure", "Geographic longitude", "Geographic and distance analysis"],
    ["Market", "str", "Dimension", "Market grouping", "Market performance"],
    ["Order City", "str", "Dimension", "Order destination city", "Geographic analysis"],
    ["Order Country", "str", "Dimension", "Order destination country", "Geographic analysis"],
    ["Order Customer Id", "int64", "Identifier", "Customer identifier associated with an order", "Customer analysis"],
    ["order date (DateOrders)", "str", "Date", "Date when the order was placed", "Order trend analysis"],
    ["Order Id", "int64", "Identifier", "Unique identifier for an order", "Order volume and order-level analysis"],
    ["Order Item Cardprod Id", "int64", "Identifier", "Product identifier associated with an order item", "Product analysis"],
    ["Order Item Discount", "float64", "Measure", "Discount amount applied to an order item", "Discount analysis"],
    ["Order Item Discount Rate", "float64", "Measure", "Discount rate applied to an order item", "Discount analysis"],
    ["Order Item Id", "int64", "Identifier", "Unique identifier for an order item", "Transaction-level analysis"],
    ["Order Item Product Price", "float64", "Measure", "Product price at order-item level", "Product pricing analysis"],
    ["Order Item Profit Ratio", "float64", "Measure", "Profit ratio of an order item", "Profitability analysis"],
    ["Order Item Quantity", "int64", "Measure", "Number of units ordered", "Order volume"],
    ["Sales", "float64", "Measure", "Sales amount", "Revenue analysis"],
    ["Order Item Total", "float64", "Measure", "Total value of an order item", "Revenue analysis"],
    ["Order Profit Per Order", "float64", "Measure", "Profit generated from an order", "Profitability analysis"],
    ["Order Region", "str", "Dimension", "Geographic region of the order", "Regional logistics performance"],
    ["Order State", "str", "Dimension", "Order destination state", "Regional analysis"],
    ["Order Status", "str", "Dimension", "Order status", "Order status analysis"],
    ["Order Zipcode", "float64", "Dimension", "Order destination postal code", "Geographic analysis"],
    ["Product Card Id", "int64", "Identifier", "Product identifier", "Product analysis"],
    ["Product Category Id", "int64", "Identifier", "Product category identifier associated with product", "Category analysis"],
    ["Product Description", "float64", "Metadata", "Product description field", "Not usable; completely missing"],
    ["Product Image", "str", "Metadata", "Product image URL/reference", "Not required for core analysis"],
    ["Product Name", "str", "Dimension", "Product name", "Product performance"],
    ["Product Price", "float64", "Measure", "Product selling price", "Product analysis"],
    ["Product Status", "int64", "Flag", "Product status indicator", "Product status monitoring"],
    ["shipping date (DateOrders)", "str", "Date", "Date when the order was shipped", "Delivery timeline"],
    ["Shipping Mode", "str", "Dimension", "Shipping method used for the order", "Shipping performance"],
]

data_dictionary_df = pd.DataFrame(
    data_dictionary,
    columns=[
        "Column",
        "Data Type",
        "Role",
        "Description",
        "Analytical Relevance"
    ]
)

data_dictionary_df

,Column,Data Type,Role,Description,Analytical Relevance
0,Type,str,Dimension,Payment transaction type,Payment analysis
1,Days for shipping (real),int64,Measure,Actual shipping duration in days,Delivery performance
2,Days for shipment (scheduled),int64,Measure,Scheduled shipping duration in days,Delivery benchmark
3,Benefit per order,float64,Measure,Benefit generated per order,Profitability analysis
4,Sales per customer,float64,Measure,Sales value associated with the customer/order,Customer value analysis
5,Delivery Status,str,Dimension,Delivery outcome or status,Delivery performance
6,Late_delivery_risk,int64,Flag,Indicator of late delivery risk,Delay analysis
7,Category Id,int64,Identifier,Product category identifier,Category analysis
8,Category Name,str,Dimension,Product category name,Category analysis
9,Customer City,str,Dimension,Customer city,Customer geography


### Data Dictionary Summary

The dataset contains 53 columns covering order transactions, customers, products, geography, financial metrics, delivery performance, and shipping information.

The columns were classified into identifiers, dimensions, measures, flags, dates, and metadata based on their analytical role.

Several fields are highly relevant to the project's supply chain and logistics objectives, particularly shipping duration, scheduled shipment duration, delivery status, late delivery risk, shipping mode, order region, sales, order quantity, and product/category attributes.

Several customer-identifying and credential-related fields are not required for the core analysis and will be excluded from the analytical dataset during the data preparation phase. The Product Description field is also not analytically usable because it is completely missing.

This classification will serve as the foundation for data cleaning, feature engineering, PostgreSQL table design, SQL analysis, and Power BI modeling in subsequent phases.

## 10. Business Interpretation

### 1. Dataset Structure

The dataset contains 180,519 order-item records and 53 columns covering transactions, customers, products, geography, sales, shipping, and delivery information.

The data is structured at the order-item level rather than the order level. This is supported by the presence of 180,519 unique Order Item Ids compared with 65,752 unique Order Ids.

Therefore, order-level metrics must be calculated using distinct Order Id rather than simply counting rows.

### 2. Delivery Performance

The dataset contains actual shipping duration, scheduled shipping duration, delivery status, and late delivery risk indicators.

The actual shipping duration ranges from 0 to 6 days, while scheduled shipping duration ranges from 0 to 4 days.

The dataset contains 98,977 records with a late delivery risk indicator of 1 and 81,542 records with an indicator of 0.

This indicates that delivery performance is a major analytical area and provides sufficient information to evaluate on-time delivery, delay rate, and shipping efficiency.

### 3. Regional Logistics Analysis

The dataset contains 5 markets, 23 order regions, 164 countries, 1,089 states, and 3,597 cities.

This geographic coverage provides sufficient granularity to compare logistics performance across different regions and identify areas with relatively high delivery delays or longer shipping times.

Regional analysis will therefore be an important component of the supply chain performance evaluation.

### 4. Shipping Mode

The dataset contains four shipping modes: Standard Class, First Class, Second Class, and Same Day.

The availability of both shipping mode and delivery performance variables allows the analysis to compare delivery time and delay rates across different shipping methods.

This can help determine whether certain shipping modes are associated with better or worse delivery performance.

### 5. Product and Category Performance

The dataset contains 50 product categories and 118 unique products.

Product and category information can be linked with delivery performance, allowing the analysis to identify categories or products that consistently experience higher delay rates or longer delivery times.

This provides an opportunity to investigate whether logistics problems are concentrated in specific product groups.

### 6. Sales and Financial Performance

The dataset contains sales, product price, discount, order quantity, and profit-related metrics.

This enables the analysis to connect logistics performance with financial performance.

For example, the project can investigate whether high-value orders experience different delivery performance compared with lower-value orders and whether operational issues may have potential financial implications.

### 7. Customer Impact

The dataset contains customer identifiers, customer segments, geographic information, and order-related information.

Although the dataset does not provide a direct customer satisfaction or customer rating variable, customer segments and order outcomes can still be used to evaluate customer-related patterns.

Customer satisfaction analysis will therefore be limited to variables available in the dataset and will not assume the existence of a direct satisfaction score.

### 8. Data Quality

The dataset contains no exact duplicate rows and no duplicate Order Item Ids.

Most analytical fields are complete. However, several missing-value issues were identified:

- Product Description is completely missing.
- Order Zipcode is missing in approximately 86.24% of records.
- Customer Lname has 8 missing values.
- Customer Zipcode has 3 missing values.

The dataset also passed the main business validation checks, including date consistency, shipping duration validity, order quantity validity, delivery status validity, customer ID consistency, and duplicate Order Item ID checks.

These findings indicate that the dataset is generally suitable for analysis, although some fields should be excluded or handled carefully during data preparation.

### 9. Outlier Interpretation

Several financial variables contain statistical outliers based on the IQR method, particularly Benefit per order, Order Profit Per Order, and Order Item Profit Ratio.

However, these observations are not automatically considered data errors because extreme values may represent legitimate high-value transactions, unusual discounts, or low-margin/high-margin orders.

Therefore, outliers will not be removed automatically. They will be evaluated based on business logic during the data cleaning and preparation phase.

### Overall Business Interpretation

The DataCo Smart Supply Chain Dataset provides sufficient information to perform a comprehensive supply chain and logistics analysis.

The dataset can support analysis of:

- Order volume
- Sales performance
- Delivery time
- Delivery delays
- Shipping mode performance
- Regional logistics performance
- Product and category performance
- Customer patterns
- Profitability

The most important analytical opportunity is to connect operational logistics performance with business outcomes.

The analysis will therefore focus not only on identifying where delays occur, but also on understanding which regions, shipping modes, products, and categories contribute most to delivery problems.

The findings from this phase will guide the data cleaning strategy, feature engineering, PostgreSQL data model, SQL analysis, exploratory data analysis, and Power BI dashboard design in the next phases.